python/pyarrow/src/arrow/python/udf.cc

```diff
@@ -28,6 +28,11 @@
 #include "arrow/util/checked_cast.h"
 #include "arrow/util/logging.h"
 
+#if PY_VERSION_HEX >= 0x030D00A4
+#define __Py_IsFinalizing Py_IsFinalizing
+#else
+#define __Py_IsFinalizing _Py_IsFinalizing
+#endif
```

Comment:
How about always using `Py_IsFinalizing` (new API)?

```suggestion
#if PY_VERSION_HEX < 0x030D00A4
#define Py_IsFinalizing() _Py_IsFinalizing()
#endif
```

Reply:
Fair, I'll merge this, and then amend in the rest of the changes.